### ISSUE TITLE
What's the use of exchangeExecutionHandler?  

### ISSUE THREAD

**Issue by JzpWh** [eclipse/californium#972]
https://github.com/eclipse/californium/blob/7b9311280a5a93dde0cef9c4c1170b061bac9f4e/californium-core/src/main/java/org/eclipse/californium/core/network/CoapEndpoint.java#L327-L339

The exchangeExecutionHandler do nothing.
Why just  use CoapEndpoint.executor to construct TcpMatcher?
https://github.com/eclipse/californium/blob/7b9311280a5a93dde0cef9c4c1170b061bac9f4e/californium-core/src/main/java/org/eclipse/californium/core/network/CoapEndpoint.java#L346-L357

**Comment by boaks** [comment]
The `CoapEndpoint` has a life-cycle:
- create (including the matcher and coap-stack)
- (optional) set executor
- start
...

With that, it's impossible to create the matcher and provide the executor, which may be provided in the later step. That "delegation executor" allows to create the matcher and coap-stack at "create", passing in that "delegation executor" and then use the optionally provided executor to delegate the execution to.
I introduced that, when I tried to fix race conditions within the `CoapEndpoint`/`Exchange`. But that fix caused a redesign and a flood of changes. So I decided to use this delegation as work-around, not to change even more.

> Why just use CoapEndpoint.executor to construct TcpMatcher?

Because it's not set, when the `TcpMatcher` is created. It's set later, either by `setExecutor` or implicit by `start`.
